cep-cache-output.md

```diff
@@ -0,0 +1,103 @@
+<tr><td> Status </td><td> In Discussion </td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &ltw.vollprecht@gmail.com&gt; </td></tr>
+<tr><td> Created </td><td> Nov 27, 2024</td></tr>
+<tr><td> Updated </td><td> </td></tr>
+<tr><td> Discussion </td><td>  </td></tr>
+<tr><td> Implementation </td><td> rattler-build </td></tr>
+</table>
+
+## Abstract
+
+This CEP aims to define the cache output for v1 multi-output recipes.
+
+## Background
+
+Sometimes it is very useful to build some code once, and then split it into multiple build artifacts (such as shared library, header files, etc.). For this reason, `conda-build` has a special, implicit top-level build.
+
+There are many downsides to the behavior of `conda-build`: it's very implicit, hard to understand and hard to debug (for example, if an output is defined with the same name as the top-level recipe, this output will get the same requirements attached as the top-level).
+
+For the v1 spec we are attempting to formalize the workings of the "top-level" build. For this, we introduce a new top-level key `cache`, that has the same values as a regular output.
+
+## Specification
+
+The top-level `cache` key looks as follows:
+
+```yaml
+cache:
+  requirements:
+    build:
+      - ${{ compiler('c') }}
+      - cmake
+      - ninja
+    host:
+      - libzlib
+    # run: ...
+    # run_constraints: ...
+    # ignore_run_exports: ...
+  build:
+    # note: in the current implementation, all other keys for build
+    #       are also valid, but unused
+    script: build.sh
+```
+
+<details> 
+  <summary>script build.sh default discussion</summary>
+We had some debate wether the cache output should _also_ default to `build.sh` or should not have any default value for the `script`. This is still undecided.
+</details>
+
+When computing variants and used variables, rattler-build looks at the union of a given output and the cache. That means, even if an output does not define any requirements, the cache would still add a variant for the `c_compiler`.
+
+When rattler-build executes the recipe, it will start by building the cache output that is appropriate for the current variant. This is computed by looking at all "used-variables" for the cache output and computing a "hash" for the cache. The build itself is executed in the same way as any other build.
+
+The variant keys that are injected at build time is the subset used by the cache output.
+
+When the cache build is done, the newly created files are moved outside of the `host-prefix`. Post-processing is not performed on the files beyond memoizing what files contain the `$PREFIX` (which is later replaced in binaries and text files with the actual build-prefix with a simple 1-to-1 replacement as the prefix-length are guaranteed to be equal).
+
+The cache _only_ restores files that were added to the prefix (conda-build also restored source files). Sources are re-created from scratch and no intermediate state is kept.
```

Comment:
That's actually a pretty serious constraint? Not being able to use a hot cache would kill certain recipes which need to run something on top of the `cache:` build - a classic example is building a library in the global stage, and the python bindings in the outputs (where the python bindings might otherwise rebuild the lib if not already present).

Another recent [example](https://github.com/conda-forge/llvmdev-feedstock/pull/297) would not work (though we decided against doing that an unrelated reason, i.e. that installing clobbered files is not guaranteed to follow the dependency relationship of outputs).

I have several other feedstocks that are installing from the intermediate build in all of the outputs, and it's unmaintainable to move these to simple `files:` directives under the current [circumstances](https://github.com/conda/conda-build/issues/5455).

Reply:
So my vision for this was to have the build folder of cmake / meson etc. in a common place and we could add a new build time variable that would be constant (e.g. `$SCRATCH`).

Then your cmake invocation could look like: 

```
cd $SCRATCH
cmake $SRC_DIR -DCMAKE_INSTALL_PREFIX=$PREFIX
ninja install
```

The $SCRATCH dir could be on `/tmp` or in the work dir, but just not part of the `$SRC_DIR`, to keep it pristine.

I tried this in [this repo](https://github.com/wolfv/rattler-build-cache-test) and ran into some issue: we do create independent build folders per output that have different names. And CMake detects that and then complains, when invoked with differnet options, that the source dir doesn't match anymore:

```txt
 │ │ + cmake $SRC_DIR -GNinja -DCMAKE_INSTALL_PREFIX=$PREFIX -DBUILD_PYTHON_BINDINGS=ON
 │ │ CMake Error: The source "$SRC_DIR/CMakeLists.txt" does not match the source "/Users/wolfv/Programs/calculator/output/bld/rattler-build_libcalculator_1732782319/work/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.
```

We have a few options:

- Keep the same path for the "work dir" the for all outputs (this will prevent concurrent builds in the future). This does not need to imply that we restore the (dirty) work dir from the cache.
- We could also restore the "dirty" work dir from the cache. For that, we would need to keep a pristine copy of the work dir at cache time, because every output might modify or add files to the work dir. In this case, we would have to prevent every output to restore the global sources, and instead only restore the dirty work dir which already has the global sources. We would also have to keep the path to the workdir the same.

We could still have parallel builds or actions if the outputs are merely splitting the existing cache files (we could find if they have a build script or not).

Reply:
All that sounds reasonable to me. I think I'd prefer option 2, but I don't understand though why 

> we would have to prevent every output to restore the global sources, and instead only restore the dirty work dir which already has the global sources.

is necessary. Some output may genuinely need to reconfigure something that needs the sources, even though they want to use a hot cache of the dirty work dir.

Reply:
This is unexpected. cache only caches files that are installed to $PREFIX? If you want to cache build artifacts, then you need to store them in $SCRATCH?

What about projects that do build-time source-code generation in-place? For example, cython projects which convert .pyx files into .c files. Or projects like magma which at build time duplicate the entire source from real float implementation into double and complex types (they don't use C++ templates). These generated sources will be need to be regenerated if the outputs need to reuse them (e.g. if you are trying to reuse the cache to amortize work across builds for multiple python versions)?